### ISSUE TITLE
Interpolate particle emit positions when emitter is moving fast

### ISSUE THREAD

**Issue by aXu-AP** [godotengine/godot-proposals#8088]
### Describe the project you are working on

Currently just helping other people learning Godot, but this is fairly common issue in my past projects as well :)

### Describe the problem or limitation you are having in your project

When particle emitter moves fast or framerate is low, particles get emitted only at discrete positions every frame, even if the amount of particles is higher, they just spawn at the same point whenever frame is rendered. Games tend to have particle emitting fast moving things (projectiles, vehicles, dashing etc.) so this is a common problem.
It is also a question which has been raised multiple times and answered sometimes incorrectly ("just ramp up the amount!"): See some examples: [reddit1](https://www.reddit.com/r/godot/comments/1760wrs/maybe_a_little_bit_silly_question_but_how_can_i/), [reddit 2](https://www.reddit.com/r/godot/comments/czxwxj/how_would_you_create_a_continuous_particle_trail/), [qa](https://ask.godotengine.org/70201/emit-particle-trail-continuously-with-fast-movement-2d),  godotengine/godot#47973.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Instead of emitting particles at the current position of the emitter (+ emission shape), interpolate the position between frames. Possibly enable the feature only if emitting frequency surpasses framerate.
I don't think this is a breaking change or why this should have a toggle. Maybe performance concerns? Feel free to express your opinion on this. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here's proof of concept screencap, code and example project:

![fire](https://github.com/godotengine/godot-proposals/assets/1621768/55d7ddae-f4c7-4443-95c1-18b1b244ee1b)
In the image, both particle emitters have the same settings, but the upper one uses custom code from below.

```
extends GPUParticles3D

@export var enabled : bool:
	set(value):
		enabled = value
		emit_particle(global_transform, Vector3.ZERO, Color.WHITE, Color.WHITE, 0b11111)
		last_emitted = Time.get_ticks_usec()
	get:
		return enabled
var old_transform : Transform3D
var old_transform_time : int
var last_emitted : int

func _process(delta: float) -> void:
	if not enabled:
		return
	var current_time = Time.get_ticks_usec()
	if not old_transform:
		old_transform = global_transform
		old_transform_time = current_time - 1
	var emit_time = lifetime * 1_000_000 / amount
	while last_emitted + emit_time < current_time:
		last_emitted += emit_time
		var w = inverse_lerp(old_transform_time, current_time, last_emitted)
		emit_particle(old_transform.interpolate_with(global_transform, w), Vector3.ZERO, Color.WHITE, Color.WHITE, 0b11111)
	old_transform = global_transform
	old_transform_time = current_time
```
[ParticleExample.zip](https://github.com/godotengine/godot-proposals/files/12886078/ParticleExample.zip)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Used often.
Some workarounds exists, but they compromise some of the quality:
* Custom emitting as demonstrated above. This has a few limitations: particles emitted from script lag behind one frame (I'm not sure if this is a bug or if it's workable), user needs to reimplement emitter shapes, initial velocities etc.
* Using bigger emitter shape or add random velocity to spread out the particles. Might or might not align with artistic intent.

### Is there a reason why this should be core and not an add-on in the asset library?

Enhances core functionality, isn't very complex, workarounds aren't perfect.

**Comment by Calinou** [comment]
cc @QbieShay

**Comment by jcostello** [comment]
Can you use the velocity to calculate the emission?

**Comment by aXu-AP** [comment]
@jcostello what are you referring to? If you are referring to "_particles emitted from script lag behind one frame_", then yes, you could predict the next position as long as the emitter is following CharacterBody or RigidBody or some custom node with exposed linear and angular velocity. Or just extrapolate current trajectory. Until it hits a wall, in which case it will overshoot. Also makes more burden for particle emission code, needs to handle different cases.

**Comment by jordo** [comment]
I have done this in other engines.... and you basically have to hook into the actual particle vertex emission positions.  Instead of generating vertex geometry at the current particle node position on any given frame, you lerp all the new particles on a line from the current particle node position to the last node position.  For GPU particles this last position needs to be sent as well to the shader so the shader can lerp between current emission position and last emission position if it needs to emit X new particles for any given frame.

**Comment by jcostello** [comment]
@aXu-AP Yes I was referring to that

**Comment by aXu-AP** [comment]
@jcostello Yeah, as you can see, there's the problem that you cannot know if the current velocity predicts correctly the next position.
@jordo Isn't this exactly what I am suggesting and what my example code does (in Godot the data is sent to gpu via `emit_particle()`)?

**Comment by jordo** [comment]
> @jordo Isn't this exactly what I am suggesting and what my example code does (in Godot the data is sent to gpu via `emit_particle()`)?

Well yes kinda.  Yes in the fact that you are emitting particles manually yourself by populating the particle buffer, but this doesn't support the particle emission shape for example, etc...  But correct me if I'm wrong, but I believe the proposal you are opening is support to have this done automagically by the particle system itself.   Perhaps you can test and implement it quickly by modifying the particle process material shader...  You really just want to lerp the emission base point, and then do the regular emission point calculations (taking emission shape into account for ex) and everything after the fact once the emission base position has been lerped.

**Comment by aXu-AP** [comment]
> You really just want to lerp the emission base point, and then do the regular emission point calculations

Ah yes, this is what I envisioned. I just wanted to make the poc short and so it consists of the very minimum needed to show the difference. I might take a look at the particle process material, I've never actually looked at the code itself.

**Comment by QbieShay** [comment]
Note that after the latest particle updates, it should't be too complicated to combine the current transform of the emitter with the velocity from the emitter, which is something that gets passed to the shader.

**Comment by aXu-AP** [comment]
@QbieShay I've totally missed that we can now inherit emitter speed to particles, that sounds wonderful! So the position could be calculated as `TRANSFORM[3].xyz += EMITTER_VELOCITY * negative_delta`?
But I guess we still would need to pass a delta value to the shader: `negative_delta = supposed_spawn_time - current_time`. Or is this info decipherable from some available data?

**Comment by jcostello** [comment]
> @QbieShay I've totally missed that we can now inherit emitter speed to particles, that sounds wonderful! So the position could be calculated as `TRANSFORM[3].xyz += EMITTER_VELOCITY * negative_delta`? But I guess we still would need to pass a delta value to the shader: `negative_delta = supposed_spawn_time - current_time`. Or is this info decipherable from some available data?

That would be awesome if it works

**Comment by OhiraKyou** [comment]
This clumping behavior limits the quality of individual particle trails quite a lot.

## Use cases

My use cases include the following:

- Bubble trails attached to fish and the player character when moving quickly underwater
- Sparks and smoke puffs coming from flying objects

## Workaround

Another workaround is to pack multiple particles into one by using one of the following:

- Textures with extra detail (e.g., a cluster of bubbles instead of a single bubble per texture)
- Meshes with extra geometry (e.g., multiple bubble spheres in a single mesh)

However, this solution adds bulk, making it incompatible with particle streams that are intended to appear thin. It also doesn't work when particles need to be scaled separately.

**Comment by bcmpinc** [comment]
I ran into this issue as well and found that `TRANSFORM[3].xyz -= EMITTER_VELOCITY * DELTA` and enabling `fract_delta`  worked well here.  Note the `-=`, using addition still causes some clumping. Disabling the fixed FPS might help reduce clumping too.
